subt/cloudsim2osgar.py

```diff
@@ -260,6 +266,20 @@ def depth_rear(self, msg):
         rospy.loginfo_throttle(10, "depth_rear callback: {}".format(self.depth_rear_count))
         self.bus.publish('depth_rear', self.convert_depth(msg))
 
+    def convert_points(self, msg):
+        # accept only Velodyne VLC-16 (for the ver0)
+        assert msg.height == 16, msg.height
+        assert msg.width == 10000, msg.width
+        assert msg.point_step == 32, msg.point_step
+        assert msg.row_step == 320000, msg.row_step
+        arr = np.frombuffer(msg.data, dtype=np.float64)
```

Comment:
just note about the type:
```
[name: "x"
offset: 0
datatype: 7
count: 1, name: "y"
offset: 4
datatype: 7
count: 1, name: "z"
offset: 8
datatype: 7
count: 1, name: "intensity"
offset: 16
datatype: 7
count: 1, name: "ring"
offset: 24
datatype: 4
count: 1]
```

Reply:
http://docs.ros.org/en/melodic/api/sensor_msgs/html/msg/PointField.html
We could handle it with https://numpy.org/doc/stable/user/basics.rec.html but for now converting it to 8 float32-typed values and using only the first 3, will do.

Reply:
It really has float32 data type: https://github.com/ignitionrobotics/ros_ign/blob/94a719adcf6005fae657a174489cb99349c33ced/ros_ign_point_cloud/src/point_cloud.cc#L348

Here is some more code for reference: https://github.com/ros/common_msgs/blob/noetic-devel/sensor_msgs/src/sensor_msgs/point_cloud2.py 

The problem in the ign code is, that it creates a point cloud message that can contain all of depth camera, rgbd camera and 3d lidar. The point_size is the maximum of the point sizes (I guess):

https://github.com/ignitionrobotics/ros_ign/blob/94a719adcf6005fae657a174489cb99349c33ced/ros_ign_point_cloud/src/point_cloud.cc#L368-L370

https://github.com/ros/common_msgs/blob/ea6c65e8846d9873b8d6e44435c52fb43d6093a3/sensor_msgs/include/sensor_msgs/impl/point_cloud2_iterator.h#L198-L229

> The fields are given as
>  * strings: "xyz" (3 floats), "rgb" (3 uchar stacked in a float),
>  * "rgba" (4 uchar stacked in a float)

So that is why we have 4 floats. I have no idea where "intensity" and "ring" come from.

Reply:
I've found it: https://github.com/ignitionrobotics/ign-sensors/blob/f8366f299947ccf9e8e39b16b9ff23655d06d521/src/GpuLidarSensor.cc#L106-L115. The "rgb" and "xyz" fields are considered a single fields of the size 12 so their memory alignment is 16.